sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsGdsStore.scala

```diff
@@ -230,26 +236,62 @@ class RapidsGdsStore(
     class RapidsGdsBatchedBuffer(
         id: RapidsBufferId,
         path: File,
-        fileOffset: Long,
-        size: Long,
+        val fileOffset: Long,
+        override val size: Long,
         meta: TableMeta,
         spillPriority: Long,
         spillCallback: RapidsBuffer.SpillCallback,
         var isPending: Boolean = true)
         extends RapidsGdsBuffer(id, size, meta, spillPriority, spillCallback) {
 
+      private def batchUnspill(buffer: DeviceMemoryBuffer): Unit = {
+        withResource(buffer) { _ =>
+          spilledBuffers.get(path).foreach { id =>
+            withResource(catalog.acquireBuffer(id)) {
+              case gdsBuffer: RapidsGdsBatchedBuffer =>
+                closeOnExcept(DeviceMemoryBuffer.allocate(gdsBuffer.size)) { d =>
+                  logDebug(s"Unspilling $gdsBuffer $id to $DEVICE")
+                  d.copyFromMemoryBuffer(
+                    0, buffer, gdsBuffer.fileOffset, gdsBuffer.size, Cuda.DEFAULT_STREAM)
+                  try {
+                    deviceStorage.copyBuffer(gdsBuffer, d, Cuda.DEFAULT_STREAM)
+                  } catch {
+                    case _: DuplicateBufferException =>
+                      logDebug(s"Lost device buffer registration race for buffer $id, ignored")
+                  }
+                }
+              case _ =>
+                logDebug(s"Ignoring buffer $id")
+            }
+          }
+        }
+      }
+
       override def materializeMemoryBuffer: MemoryBuffer = this.synchronized {
-        closeOnExcept(DeviceMemoryBuffer.allocate(size)) { buffer =>
-          if (isPending) {
+        if (isPending) {
+          closeOnExcept(DeviceMemoryBuffer.allocate(size)) { buffer =>
             copyToBuffer(buffer, fileOffset, size, Cuda.DEFAULT_STREAM)
             Cuda.DEFAULT_STREAM.sync()
             logDebug(s"Created device buffer $size from batch write buffer")
-          } else {
+            buffer
+          }
+        } else if (RapidsBufferCatalog.shouldUnspill) {
+          closeOnExcept(DeviceMemoryBuffer.allocate(batchWriteBufferSize)) { buffer =>
+            CuFile.readFileToDeviceMemory(buffer.getAddress, buffer.getLength, path, 0)
+            logDebug(s"Created device buffer for $path 0:$batchWriteBufferSize via GDS")
+            closeOnExcept(DeviceMemoryBuffer.allocate(size)) { d =>
+              d.copyFromMemoryBuffer(0, buffer, fileOffset, size, Cuda.DEFAULT_STREAM)
```

Comment:
so there's a copy here, but there's also another one in the thread. It's not clear what this copy is, and what the other copy is within the thread. Just by looking at this diff. Perhaps I am missing something.

Reply:
This copy is returned right away, while the whole batched buffer is unspilled in the background thread.